### PR TITLE
Add support for generating property schema format for Url and Hostname

### PR DESCRIPTION
**Pull request by norkunas**
| Q             | A
| ------------- | ---
| Branch?       | main
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | N/A
| License       | MIT
| Doc PR        | N/A

Add a `PropertySchemaUrlRestriction` to transform [Url](https://symfony.com/doc/current/reference/constraints/Url.html) validation constraint into json schema.

### CODE REVIEW

**Comment by guilliamxavier** [comment]
Please remove me from the description :laughing: I'm already checking https://github.com/api-platform/core/pulse from time to time these days (also I'm not part of the team)

If it may help: the test failure for "lowest" is related to https://github.com/symfony/validator/commit/adab2122723e3f59cc091446999138ff83825ffb (introduced in v4.4.8)

But string manipulation and generating a huge pattern in the schema... is it really worth it? 🤔 Wouldn't it be enough to add something like this in the existing `PropertySchemaFormat`?

```diff
         if ($constraint instanceof Email) {
             return ['format' => 'email'];
         }
+
+        if ($constraint instanceof Url) {
+            return ['format' => 'uri'];
+        }
 
         if ($constraint instanceof Uuid) {
             return ['format' => 'uuid'];
         }
```

This would not be strictly equivalent (also ambiguity between "URL", "URI" and "URI reference"), but that's already true for `Email` (3 different validation modes)... Also `Url` is probably less used...

Anyway let's wait for a proper review from a team member :smile:

**Comment by norkunas** [comment]
I just cc'ed because you were interested in these PR's :D

**Comment by norkunas** [comment]
IMHO if backend validates based on some pattern then it should be exposed nevertheless..

**Comment by norkunas** [comment]
Included also Hostname constraint as it's uri related

**Comment by alanpoulain** [comment]
I agree with @guilliamxavier. Since the format is used, I don't think it's a good idea to have the regexp as well.
Since URI is a specification, there should be no difference between the "client" regular expression and the backend one.

**Comment by norkunas** [comment]
But how would you distinguish then allowing only some other specific protocols?

**Comment by alanpoulain** [comment]
You would not be able to, that's right. Personally, I think it's a little bit too much detailed though. And the implementation is very dependent of the pattern in `UrlValidator`. If something change, it will break.
But maybe I'm wrong on this one. @soyuka, do you have an opinion on this matter?

File: src/Bridge/Symfony/Validator/Metadata/Property/Restriction/PropertySchemaUrlRestriction.php
Comment (soyuka) [inline src/Bridge/Symfony/Validator/Metadata/Property/Restriction/PropertySchemaUrlRestriction.php:49]: I'd suggest to use the available `uri` format on string hence:

```
[
  'type' => 'string',
  'format' => 'uri'
]
```

There are many `URL` formats out there and many corresponding formats on the json-schema specification (http://json-schema.org/draft/2020-12/json-schema-validation.html#rfc.section.7.3). Might be that Url is actually too broad here and it would need the user to specify the format himself?

Using `#[ApiProperty(schema: ['format' => 'uri'])]` should work (needs check).

Comment (norkunas) [reply src/Bridge/Symfony/Validator/Metadata/Property/Restriction/PropertySchemaUrlRestriction.php:49]: > I don't get the use of a pattern here, json-schema already validates uris. I'd suggest to use the available `uri` format on string hence:
> 
> ```
> [
>   'type' => 'string',
>   'format' => 'uri'
> ]
> ```
> 
> There are many `URL` formats out there and many corresponding formats on the json-schema specification (http://json-schema.org/draft/2020-12/json-schema-validation.html#rfc.section.7.3). Might be that Url is actually too broad here and it would need the user to specify the format himself?
> 
> Using `#[ApiProperty(schema: ['format' => 'uri'])]` should work (needs check).

I've added uri format to PropertySchemaFormat

Comment (soyuka) [reply src/Bridge/Symfony/Validator/Metadata/Property/Restriction/PropertySchemaUrlRestriction.php:49]: okay got it sorry I missed this part, you don't need the regex, it's the responsibility of the schema validator to validate an URI not ours in my opinion.

Comment (norkunas) [reply src/Bridge/Symfony/Validator/Metadata/Property/Restriction/PropertySchemaUrlRestriction.php:49]: Okay will remove that part

**Comment by norkunas** [comment]
Removed the pattern part.

**Review by soyuka — APPROVED** [review]

**Comment by soyuka** [comment]
Tyvm @norkunas !

**Pull request merged**
